FIAT/brezzi_douglas_marini_cube.py

```diff
@@ -0,0 +1,229 @@
+
+from sympy import symbols, legendre, Array, diff
+import numpy as np
+from FIAT.finite_element import FiniteElement
+from FIAT.dual_set import make_entity_closure_ids
+from FIAT.polynomial_set import mis
+from FIAT.reference_element import compute_unflattening_map, flatten_reference_cube
+
+x, y, z = symbols('x y z')
+variables = (x, y, z)
+leg = legendre
+
+
+def triangular_number(n):
+    return int((n+1)*n/2)
+
+
+class BrezziDouglasMariniCube(FiniteElement):
+    def __init__(self, ref_el, degree, mapping):
+        if degree < 1:
+            raise Exception("BDMce_k elements only valid for k >= 1")
+
+        flat_el = flatten_reference_cube(ref_el)
+        dim = flat_el.get_spatial_dimension()
+        if dim != 2:
+            raise Exception("BDMce_k elements only valid for dimension 2")
+
+        flat_topology = flat_el.get_topology()
+
+        entity_ids = {}
+        cur = 0
+
+        for top_dim, entities in flat_topology.items():
+            entity_ids[top_dim] = {}
+            for entity in entities:
+                entity_ids[top_dim][entity] = []
+
+        for j in sorted(flat_topology[1]):
+            entity_ids[1][j] = list(range(cur, cur + degree + 1))
+            cur = cur + degree + 1
+
+        entity_ids[2][0] = list(range(cur, cur + 2*triangular_number(degree - 1)))
+        cur += 2*triangular_number(degree - 1)
+
+        formdegree = 1
+
+        entity_closure_ids = make_entity_closure_ids(flat_el, entity_ids)
+
+        super(BrezziDouglasMariniCube, self).__init__(ref_el=ref_el, dual=None, order=degree, formdegree=formdegree,
+                                                      mapping=mapping)
+
+        topology = ref_el.get_topology()
+        unflattening_map = compute_unflattening_map(topology)
+        unflattened_entity_ids = {}
+        unflattened_entity_closure_ids = {}
+
+        for dim, entities in sorted(topology.items()):
+            unflattened_entity_ids[dim] = {}
+            unflattened_entity_closure_ids[dim] = {}
+        for dim, entities in sorted(flat_topology.items()):
+            for entity in entities:
+                unflat_dim, unflat_entity = unflattening_map[(dim, entity)]
+                unflattened_entity_ids[unflat_dim][unflat_entity] = entity_ids[dim][entity]
+                unflattened_entity_closure_ids[unflat_dim][unflat_entity] = entity_closure_ids[dim][entity]
+        self.entity_ids = unflattened_entity_ids
+        self.entity_closure_ids = unflattened_entity_closure_ids
+        self._degree = degree
+        self.flat_el = flat_el
+
+    def degree(self):
+        return self._degree
+
+    def get_nodal_basis(self):
+        raise NotImplementedError("get_nodal_basis not implemented for bdmce")
+
+    def get_dual_set(self):
+        raise NotImplementedError("get_dual_set is not implemented for bdmce")
+
+    def get_coeffs(self):
+        raise NotImplementedError("get_coeffs not implemented for bdmce")
```

Comment:
Even though the construction is done on the prime basis directly (for tabulation), it would be good to implement a dual basis given that these elements _are_ interpolatory.